libmattermost.c

```diff
@@ -1283,53 +1497,98 @@ mm_get_info(PurpleConnection *pc,const gchar *username)
 	} \
 }
 
+static void mm_refresh_statuses(MattermostAccount *ma, const gchar *id);
+
+int mm_compare_users_by_username_int(gconstpointer a, gconstpointer b)
+{
+	const MattermostUser *u1 = a;
+	const MattermostUser *u2 = b;
+
+	gint res = g_strcmp0(u1->username, u2->username);
+
+	if (res > 0) { return 1; }
+	if (res < 0) { return -1; }
+	return 0;
+}
+
+
 static void
 mm_get_users_by_ids_response(MattermostAccount *ma, JsonNode *node, gpointer user_data)
 {
 	JsonObject *response = json_node_get_object(node);
-	PurpleGroup *default_group = mm_get_or_create_default_group();
-	MattermostUser *mm_user;	
-	GList *i;
 
-	for (i=user_data;i;i=i->next) {
+	if (json_object_get_int_member(response, "status_code") >= 400) {
+		// do not report error to UI: may be called from websocket callback
+		// TODO: improve
+		return;
+	}
+
+	PurpleGroup *default_group = mm_get_or_create_default_group();
+	MattermostUser *mm_user;
+	GList *mm_users = user_data;	
+	GList *i = NULL;
+	
+	for (i=mm_users;i;i=i->next) {
 		mm_user = i->data;
 		JsonObject *user = json_object_get_object_member(response,mm_user->user_id);
 		if (user != NULL) {			
-			const gchar *username = json_object_get_string_member(user, "username");
-			if (username != NULL) {          		
-				PurpleBuddy *buddy = purple_buddy_new(ma->account, username, NULL);
-				purple_blist_add_buddy(buddy, NULL, default_group, NULL);
-				g_hash_table_replace(ma->one_to_ones, g_strdup(mm_user->user_id), g_strdup(username));
-				g_hash_table_replace(ma->one_to_ones_rev, g_strdup(username), g_strdup(mm_user->user_id));
-
-				_MM_BLIST_SET(buddy,user,"room_id",mm_user->room_id);
-				_MM_BLIST_SET(buddy,user,"user_id",mm_user->user_id);
-				_MM_BLIST_SET(buddy,user,"nickname",NULL);
-				_MM_BLIST_SET(buddy,user,"first_name",NULL);
-				_MM_BLIST_SET(buddy,user,"last_name",NULL);
-				_MM_BLIST_SET(buddy,user,"email",NULL);
-
-				mm_get_avatar(ma,buddy);
-			}
-
-		g_free(mm_user->user_id);
-		g_free(mm_user->room_id);
-		g_free(mm_user);
+			mm_user->username = g_strdup(json_object_get_string_member(user, "username"));
+			mm_user->nickname = g_strdup(json_object_get_string_member(user, "nickname"));
+			mm_user->first_name = g_strdup(json_object_get_string_member(user, "first_name"));
+			mm_user->last_name = g_strdup(json_object_get_string_member(user, "last_name"));
+			mm_user->email = g_strdup(json_object_get_string_member(user, "email"));
 		}
 	}
 
-	g_list_free(user_data);
+	mm_users = g_list_sort(mm_users, mm_compare_users_by_username_int);

```

Comment:
code below ?  yes: without sort buddies will be created in random order, with in username alphabetical order: since I already made channels appear in alphabetical order thought that this will nicely fit together.